Refactor attachment handling and UI components

w15eacre opened the pull request:
Refactor attachment editor to support editing files.

- Renamed NewEntryAttachmentsDialog to EditEntryAttachmentsDialog for clarity.
- Introduced EditEntryAttachmentsDialog class to manage editing of existing attachments.
- Added functionality to edit attachments in EntryAttachmentsWidget.
- Created AbstractAttachmentWidget interface for different attachment types.
- Implemented TextAttachmentsWidget for handling plain text attachments.
- Added UnknownAttachmentTypeWidget for unsupported attachment types.
- Updated PreviewEntryAttachmentsDialog to utilize the new attachment widget factory.
- Enhanced EntryAttachmentsModel with rowByKey method for better key management.
- Updated UI files for new and existing dialogs to reflect changes.
- Introduced IAttachmentWidgetFactory interface for creating attachment widgets based on MIME type.
- Add image attachment support with zoom functionality
- Fix editing a empty file
- Add the preview button on the Edit widget.
- The text preview was changed
- Add html and markdown detection 

## Screenshots
![edit_text_and_preview](https://github.com/user-attachments/assets/55af5b3a-1b7c-4f9e-b151-72a3cfaf483c)
![edit_text](https://github.com/user-attachments/assets/6ed5bcc1-9d23-4350-a5b8-761e155d297a)
![new buttons](https://github.com/user-attachments/assets/5fbdb5d0-44b6-4c9d-83e6-4204f57ba7e7)
![new_file](https://github.com/user-attachments/assets/addb7bc5-aa23-477b-afeb-a4a76a5bcd69)
![preview_image_zoom](https://github.com/user-attachments/assets/d470956d-b423-431d-b2b5-cd5c7b6b260c)
![preview html](https://github.com/user-attachments/assets/08427ad7-ae7c-4e04-959c-1109520b6ab6)

![preview_plain_text](https://github.com/user-attachments/assets/925a98cc-e256-4b95-8bb9-b6a98145bf28)
![preview_text_html](https://github.com/user-attachments/assets/de1bca40-14e8-482a-824b-47f6e97b1e71)
![Preview](https://github.com/user-attachments/assets/ab8a9127-6caf-4198-8fca-487244d0bc85)
![previw_image_fit](https://github.com/user-attachments/assets/569532d8-83e3-45e1-afde-d635bf5ee914)
![unknown_type](https://github.com/user-attachments/assets/24adf7a1-8787-4639-b3dc-b1fd19f2514e)



## Testing strategy
manually testing and unit tests


## Type of change
- ✅ New feature (change that adds functionality)
- ✅ Refactor (significant modification to existing code)

w15eacre committed:
Refactor attachment handling and UI components

- Renamed NewEntryAttachmentsDialog to EditEntryAttachmentsDialog for clarity.
- Introduced EditEntryAttachmentsDialog class to manage editing of existing attachments.
- Added functionality to edit attachments in EntryAttachmentsWidget.
- Created AbstractAttachmentWidget interface for different attachment types.
- Implemented TextAttachmentsWidget for handling plain text attachments.
- Added UnknownAttachmentTypeWidget for unsupported attachment types.
- Updated PreviewEntryAttachmentsDialog to utilize the new attachment widget factory.
- Enhanced EntryAttachmentsModel with rowByKey method for better key management.
- Updated UI files for new and existing dialogs to reflect changes.
- Introduced IAttachmentWidgetFactory interface for creating attachment widgets based on MIME type.

w15eacre commented:
Hello @droidmonkey. I've started refactoring the attachment handling code to make it easier to add new features (like PDF support and text file editing). Do we need to support rich text files as well?

droidmonkey commented:
RichText, maybe, I'm more inclined to support markdown rendering instead

w15eacre commented:
> RichText, maybe, I'm more inclined to support markdown rendering instead

I hadn’t thought about using Markdown before, but it seems interesting.

w15eacre committed:
Update translation

w15eacre commented:
@xboxones1 This PR addresses your issue. I separated the widgets by file type to enhance the editor. Now, when you press the ‘Add’ button, it generates the name “New Attachment_{index}.txt”. You can change the name and/or press the ‘Edit’ button to modify the file.

w15eacre commented:
@xboxones1 
It’s related to the MIME type. I haven’t pushed the fix yet

w15eacre committed:
Add image attachment support with zoom functionality

droidmonkey commented:
@w15eacre do you think you can have this review ready this week? I'd like to push this out with 2.7.11

w15eacre commented:
![image fit](https://github.com/user-attachments/assets/a3dcde42-e3ce-420f-b639-b7b90c411b80)
![image zoom](https://github.com/user-attachments/assets/03fba9c7-2dc5-45f4-8bee-0f0ce2ac84c5)

w15eacre commented:
![Md file](https://github.com/user-attachments/assets/dcec7c46-a37b-4b7f-a5c4-8d5486eca0df)

w15eacre commented:
![edit and preview](https://github.com/user-attachments/assets/c87fb4b3-017a-440d-9e18-3735a3f96bce)

w15eacre committed:
Adds text preview widget with support for multiple formats

Introduces a new `TextAttachmentsPreviewWidget` to enable previewing attachments in plain text, HTML, and Markdown formats.

Integrates the preview widget into `TextAttachmentsWidget` using a splitter for better attachment editing and viewing. Updates the attachment-handling logic to synchronize edits and previews dynamically.

Also expands MIME type detection by adding support for zero-size text files.

w15eacre commented:
@

> @w15eacre do you think you can have this review ready this week? I'd like to push this out with 2.7.11

I’ll finish it in 2–3 days.
Feel free to check my screen — and if you have any suggestions to improve the editor UI, just let me know.

w15eacre commented:
@xboxones1 I fixed the Edit button.

droidmonkey commented:
wow I am really into this, awesome work!

michaelk83 commented:
> if you have any suggestions to improve the editor UI

Maybe move the type dropdown to the left end of the buttons row? This way, it won't take up vertical space from the editor. And you can probably skip the label, as it's clear enough from the context. Just add it as a tooltip of the dropdown:
`[ Html      ▼ ]                                         [     OK     ] [   Cancel   ]`

Not sure if it's a good idea, but just something to consider.

w15eacre commented:
> > if you have any suggestions to improve the editor UI
> 
> 
> 
> Maybe move the type dropdown to the left end of the buttons row? This way, it won't take up vertical space from the editor. And you can probably skip the label, as it's clear enough from the context. Just add it as a tooltip of the dropdown:
> 
> `[ Html      ▼ ]                                         [     OK     ] [   Cancel   ]`
> 
> 
> 
> Not sure if it's a good idea, but just something to consider.

Not sure, but aligning the text editor and preview is a good idea

w15eacre committed:
Update signal-slot connections and improve selection handling

Replaces deprecated SIGNAL-SLOT syntax with modern signal-slot syntax for better readability and maintenance.

Enhances attachment selection handling by adding a check for empty selections and providing a warning message, preventing potential crashes or undefined behavior when no attachment is selected.

w15eacre commented on src/gui/entry/EntryAttachmentsWidget.cpp:
```diff
@@ -218,6 +240,38 @@ void EntryAttachmentsWidget::previewSelectedAttachment()
+{
+    Q_ASSERT(m_entryAttachments);
+
+    const auto index = m_ui->attachmentsView->selectionModel()->selectedIndexes().first();
```
fixed

w15eacre commented on src/gui/entry/EditEntryAttachmentsDialog.cpp:
```diff
@@ -0,0 +1,75 @@
+    m_ui->dialogButtons->setStandardButtons(QDialogButtonBox::Ok | QDialogButtonBox::Cancel);
+
+    connect(m_ui->dialogButtons, SIGNAL(accepted()), this, SLOT(accept()));
+    connect(m_ui->dialogButtons, SIGNAL(rejected()), this, SLOT(reject()));
```
fixed

w15eacre committed:
Add test for MIME type detection in TestTools

w15eacre committed:
Fix include order by adding missing memory header in EditEntryAttachmentsDialog

w15eacre committed:
Switch to shared_ptr for attachment factories and improve UI logic

Replaces `unique_ptr` with `shared_ptr` for attachment widget factories to better support shared ownership and reuse across dialogs. Updates related class interfaces and member variables accordingly.

Improves attachment editing logic by handling cases where no selection is made. Refines zoom functionality for image attachments by introducing a fit-in-view factor calculation and simplifying zoom constraints.

Includes minor code cleanups and adherence to modern C++ practices.

w15eacre committed:
Fix formatting of enum declaration in TextAttachmentsPreviewWidget

w15eacre committed:
Fix include for attachment widget factory in PreviewEntryAttachmentsDialog

w15eacre committed:
Fix memory include and improve selection model checks in EntryAttachmentsWidget

w15eacre commented:
I'm trying to build the project on Ubuntu, but it fails because I'm using the setMarkdown() function. It seems that the Qt version installed on my system doesn't support this API

Do you know which Qt version KeePassXC is expected to support or target on Ubuntu? Should I avoid using setMarkdown() for compatibility reasons?

droidmonkey commented:
Wrap that call in a minimum qt version

w15eacre commented:
> Is it possible to implement automatic markup selection for the editor and viewer for text files? For example, if a document has Markdown formatting, then both during editing and viewing, it should automatically be in Markdown.
> 
> 
> 
> The preview area during editing should, by default, be visible, with the option to hide it. If I hadn't seen screenshots, I wouldn't have immediately understood that it was there. Or there should be an option to change this default behavior in the settings


We can’t see the difference between these sorts of text by MIME type, because it’s just text/plain. But we can check the text and try to find special symbols.

I’m going to add a Preview button on the editor page to show or hide the preview widget.

michaelk83 commented:
> We can’t see the difference between these sorts of text by MIME type, because it’s just text/plain.

There are specific MIME types for Markdown, HTML, etc. text/plain is the wrong MIME type for those.

droidmonkey commented:
MIME is not stored with a file. It is either inferred, detectable (like a magic number), or explicitly selected. So you always start with text/plain and if it becomes obvious that its actually HTML or Markdown through inference than great. The kdbx standard doesn't offer the ability to store mime type with files either which sucks.

w15eacre commented:
> > We can’t see the difference between these sorts of text by MIME type, because it’s just text/plain.
> 
> 
> 
> There are specific MIME types for Markdown, HTML, etc. text/plain is the wrong MIME type for those.

@droidmonkey answered you. I’ll add some additional information. We use QMimeDatabase, which extracts information from raw data, but it can’t always determine the correct MIME type.

I wrote a test that uses different headers to get different MIME types. However, when the input is text, it always returns text/plain, and we have to guess what kind of text it actually is.

w15eacre committed:
Add TextAttachmentsEditWidget and integrate with TextAttachmentsWidget

w15eacre committed:
Remove markdown if the QT version < 5.14

w15eacre committed:
Add translation context for TextAttachmentsEditWidget

w15eacre committed:
Remove unused "Form" title from multiple attachment widgets and update text format in UnknownAttachmentTypeWidget

w15eacre committed:
Enhance MIME type handling for attachments

- Refactor toMimeType function to support HTML and Markdown formats.
- Update getMimeType to handle QFileInfo for better file type detection.
- Modify attachment widget factory to create widgets for HTML and Markdown types.
- Improve button enabling logic in EntryAttachmentsWidget based on MIME type.
- Add tests for new MIME type handling and file info detection.

w15eacre commented:
@michaelk83 I have added markdown and html detection by a file extention

w15eacre commented:
> Maybe add an edit button to the preview panel as well? 

Hm...

w15eacre committed:
Prevent collapsing of the edit widget in TextAttachmentsWidget

w15eacre committed:
Add tests:

- AttachmentsWidgetFactory
- PreviewEntryAttachmentsDialog

w15eacre commented:
@droidmonkey I’ve started adding tests

w15eacre commented on src/gui/entry/EntryAttachmentsWidget.cpp:
```diff
@@ -346,35 +418,40 @@
 void EntryAttachmentsWidget::updateButtonsEnabled()
 {
-    const bool hasSelection = m_ui->attachmentsView->selectionModel()->hasSelection();
+    const auto selectionModel = m_ui->attachmentsView->selectionModel();
```
Fixed

w15eacre committed:
Refactors attachment handling and improves QString usage

Replaces repetitive inline `attachment.name` references with a
dedicated `QString` variable for clarity and reusability.

Uses `std::move` when passing attachments to improve performance.
Simplifies warning messages by removing unnecessary `tr()` calls
for non-translatable strings.

Refactors attachment handling for clarity and performance

Replaces repetitive inline `attachment.name` references with a dedicated `QString` variable to improve code clarity and reusability.

Introduces `std::move` when passing attachments to enhance performance by avoiding unnecessary copies.

Simplifies warning messages by removing `tr()` calls for non-translatable strings, aligning with best practices.

w15eacre commented:
> Maybe add an edit button to the preview panel as well? In classic KeePass, you can edit from the preview panel by clicking on the file name. It's more convenient than having to go into edit entry every time.

The current changes are sufficiently comprehensive. Should any usability issues arise (for myself or other users), I’ll prioritize implementing the requested functionality. However, I currently lack the capacity to evaluate this thoroughly.

droidmonkey commented:
Agree we can always tweak this later, this is already a rather large change and welcome change.

Issues I noticed while testing:
* Preview choices are in an odd order, should be: Plain Text, Markdown, HTML
* The preview panel incurs a major lag/delay when shown and hidden, seems like we might have an inefficiency here. I tested this with an 800KB XML file, all preview options incur a lag. This is especially noticeable when the preview panel is hidden. After doing CPU profiling this appears to be entirely a Qt issue, and may go away when using release libraries. Either way we may be doing unnecessary rendering when hiding the preview panel, something to look at.
* Previewing Plain Text when editing an attachment is not necessary
* Double clicking a row doesn't do anything. Recommend double click when editing an entry opens the "Edit Attachment" and double click when previewing an entry opens "Preview Attachment". If an attachment type cannot be edited then double click opens the preview.
* The difference between "New" and "Add" is not clear anymore. Perhaps change "Add" to "Browse..."? or "Add File..."
* After clicking "New", editing the file name, then pressing "Enter" the attachment is not selected anymore and you have to find it in the list, select it, then click edit. Perhaps after setting the name then pressing enter, the editing widget would open up immediately for that attachment. The other option is to not add a row at all and set the name of the attachment in the editing widget instead.


After looking at the code, I find it to be needlessly complex. We don't need a factory to produce 3 different types of widgets. The orchestration of the preview widget is unclear in the code flow since it bounces signals around the various widgets. I think this can be greatly simplified.
* BaseAttachmentWidget -> Forms the frame for the below two. Default is to show "Cannot preview" in the middle which is then replaced by Text or Image inserts if it can be previews.
* TextAttachmentWidget -> Hides preview button if not editing. Combines Preview and Edit widgets.
* ImageAttachmentWidget -> Nothing to change here

Basically make the BaseAttachmentWidget the overarching container that can then insert specific preview widgets depending on the attachment or a "No Preview Possible" message. It also changes the button box arrangement depending on Edit vs Preview.

w15eacre committed:
Add unit tests for attachment-related GUI components

Introduces comprehensive unit tests for various attachment-related GUI components, including dialogs, widgets, and views. Enhances test coverage to ensure functionality such as high DPI scaling, attachment handling, zoom controls, and widget-specific behaviors.

Improves maintainability and reliability by verifying expected behaviors, signal emissions, and edge cases in the GUI components.

Relates to improving attachment feature robustness and ensuring compatibility with different attachment types.

w15eacre commented:
> Agree we can always tweak this later, this is already a rather large change and welcome change.
> 
> Issues I noticed while testing:
> 
>     * Preview choices are in an odd order, should be: Plain Text, Markdown, HTML
> 
>     * The preview panel incurs a major lag/delay when shown and hidden, seems like we might have an inefficiency here. I tested this with an 800KB XML file, all preview options incur a lag. This is especially noticeable when the preview panel is hidden. After doing CPU profiling this appears to be entirely a Qt issue, and may go away when using release libraries. Either way we may be doing unnecessary rendering when hiding the preview panel, something to look at.
> 
>     * Previewing Plain Text when editing an attachment is not necessary
> 
>     * Double clicking a row doesn't do anything. Recommend double click when editing an entry opens the "Edit Attachment" and double click when previewing an entry opens "Preview Attachment". If an attachment type cannot be edited then double click opens the preview.
> 
>     * The difference between "New" and "Add" is not clear anymore. Perhaps change "Add" to "Browse..."? or "Add File..."
> 
>     * After clicking "New", editing the file name, then pressing "Enter" the attachment is not selected anymore and you have to find it in the list, select it, then click edit. Perhaps after setting the name then pressing enter, the editing widget would open up immediately for that attachment. The other option is to not add a row at all and set the name of the attachment in the editing widget instead.
> 
> 
> After looking at the code, I find it to be needlessly complex. We don't need a factory to produce 3 different types of widgets. The orchestration of the preview widget is unclear in the code flow since it bounces signals around the various widgets. I think this can be greatly simplified.
> 
>     * BaseAttachmentWidget -> Forms the frame for the below two. Default is to show "Cannot preview" in the middle which is then replaced by Text or Image inserts if it can be previews.
> 
>     * TextAttachmentWidget -> Hides preview button if not editing. Combines Preview and Edit widgets.
> 
>     * ImageAttachmentWidget -> Nothing to change here
> 
> 
> Basically make the BaseAttachmentWidget the overarching container that can then insert specific preview widgets depending on the attachment or a "No Preview Possible" message. It also changes the button box arrangement depending on Edit vs Preview.

I got your feedback about behavior and performance refactoring — no problem, I will take care of it.
Now, let's talk about the architecture refactoring.

I use the factory to delegate the responsibility of choosing the widget to another class, in order to avoid overriding the dialogs. For example, if I want to support PDF files, I can simply create a new widget and update the factory — without having to change PreviewEntryAttachmentsDialog or EditEntryAttachmentsDialog.

If I merge TextAttachmentsEditWidget and TextAttachmentsPreviewWidget into a single TextAttachmentsWidget, I will end up with a lot of unrelated code in one class, which violates the Single Responsibility Principle.

Regarding your quote:

    "The orchestration of the preview widget is unclear in the code flow since it bounces signals around the various widgets."
Are you referring to how TextAttachmentsWidget works?

Maybe it would help to create a UML diagram to clarify the architectural concerns in my solution?

droidmonkey commented:
Please dont diagram, hah.

w15eacre commented:
Could you explain to me what the BaseAttachmentWidget is supposed to do? I have Edit and Preview dialogs — they insert the appropriate widget based on the MIME type.

Do you want the preview widget to be hidden in edit mode?

I’m trying to understand what exactly needs to be fixed in the architecture.

droidmonkey commented:
It might be easiest to describe why the factory pattern is not desired here:
1. You create three levels of indirection to achieve the same basic outcome, this code block:
```
if (type == Tools::MimeType::PlainText || type == Tools::MimeType::Html || type == Tools::MimeType::Markdown) {
    return new TextAttachmentsWidget(parent);
} else if (type == Tools::MimeType::Image) {
    return new ImageAttachmentsWidget(parent);
}

return new UnknownAttachmentTypeWidget(parent);
```
2. If you merge AbstractAttachmentWidget and UnknownAttachmentTypeWidget and the above code block into `BaseAttachmentViewWidget` then you eliminate 4 code files.
3. The BaseAttachmentViewWidget takes the input from "setAttachment" and figures out which central widget to create and ultimately display. I am ok with keeping Preview and Edit widgets separate for clarity of purpose.
4. While we are at it, you definitely don't need an `attachments` namespace
5. Also please use Qt pointer types (QSharedPointer, QScopedPointer) and not the standard library variants. We are trying to be consistent on that front.

So the flow looks like:
EntryAttachmentsWidget --creates--> [PreviewEntryAttachmentWidget, EditEntryAttachmentWidget] --creates--> BaseAttachmentViewWidget(attachment, mode) --incorporates--> [TextAttachmentViewWidget, TextAttachmentEditWidget, ImageAttachmentViewWidget]

w15eacre commented:
@droidmonkey I checked the performance issue. This is a QTextBrowser problem — we need to implement a TextPaginator.

I noticed the issue with QTextEdit/QPlainTextEdit (it also exists in the release build, though it's less noticeable than in the debug build). The solution is to implement a custom lazy text viewer, but I'm not sure this should be addressed in this PR.

w15eacre committed:
Add sorting into the TextTypeCombobox

w15eacre committed:
Enhance attachment editing by connecting QLineEdit commitData to update current index

w15eacre committed:
Refactor double-click behavior in attachments view to toggle between preview and edit modes based on read-only state

w15eacre committed:
Add context menu to add attachments in EntryAttachmentsWidget

w15eacre commented:
A new button "Add file..."
![image](https://github.com/user-attachments/assets/65be8d50-aa61-4de9-98a6-7aa88d499f4a)
![Screenshot_20250531_162003](https://github.com/user-attachments/assets/472e944e-eb1a-4f17-99af-6916b6f71116)

droidmonkey commented:
Oh I love that

w15eacre committed:
Refactor attachment handling in GUI

- Removed the AbstractAttachmentWidget class and its associated files.
- Introduced AttachmentTypes.h to define Attachment structure and OpenMode enum.
- Created AttachmentWidget class to manage attachments, replacing the abstract class.
- Updated ImageAttachmentsWidget, TextAttachmentsWidget, and related classes to inherit from QWidget instead of AbstractAttachmentWidget.
- Removed the AttachmentWidgetFactory and its tests, simplifying attachment widget creation.
- Added tests for AttachmentWidget to ensure proper functionality with text and image attachments.
- Cleaned up unused code and adjusted tests to reflect the new structure.

w15eacre committed:
Remove unused messages and update attachment dialog labels for clarity

droidmonkey commented:
Just tested the latest, this is an awesome upgrade.

droidmonkey committed:
Code cleanup

droidmonkey committed:
Functional fixes

* Fix editing attachment not saving the edits
* Fix preview being updated even if not visible and adding a delay to update to reduce typing lockups

droidmonkey commented:
@w15eacre I pushed some code standards cleanup and a couple fixes. Edited attachments were not saved (I fixed that) so might need to add a gui test for that.

w15eacre committed:
Add processEvents calls to improve UI responsiveness in attachment handling tests
Add a returning text attachment test

w15eacre commented:
@droidmonkey I’ve fixed the tests. Thanks!

droidmonkey commented on src/gui/entry/EntryAttachmentsWidget.cpp:
```diff
@@ -17,23 +17,41 @@
+    constexpr const char* DefaultName = "New Attachment";
+    constexpr const char* Suffix = ".txt";
+
+    QString generateUniqName(const QString& name, const QStringList& existingNames)
```
Should probably spell this function name out for consistently sake: `generateUniqueName`

droidmonkey commented on src/gui/entry/EntryAttachmentsWidget.cpp:
```diff
@@ -67,14 +85,31 @@ EntryAttachmentsWidget::EntryAttachmentsWidget(QWidget* parent)
+    auto addButtonMenu = new QMenu(this);
+    addButtonMenu->addAction(tr("New Text Document"), this, &EntryAttachmentsWidget::newAttachments);
+    addButtonMenu->addAction(
+        tr("Load from Disk..."), this, QOverload<>::of(&EntryAttachmentsWidget::insertAttachments));
```
Switch ... to … (ellipses character)

droidmonkey commented on src/gui/entry/EntryAttachmentsWidget.cpp:
```diff
@@ -218,6 +262,45 @@ void EntryAttachmentsWidget::previewSelectedAttachment()
+    const auto index = selectedIndexes.first();
+
+    if (!index.isValid()) {
+        qWarning() << tr("Failed to edit an attachment: Attachment not found");
```
We typically don't translate console warnings.

droidmonkey commented on src/gui/entry/EntryAttachmentsWidget.cpp:
```diff
@@ -218,6 +262,45 @@ void EntryAttachmentsWidget::previewSelectedAttachment()
+
+    const auto selectedIndexes = m_ui->attachmentsView->selectionModel()->selectedIndexes();
+    if (selectedIndexes.isEmpty()) {
+        qWarning() << tr("Failed to edit an attachment: No attachment selected");
```
We typically don't translate console warnings.

droidmonkey commented on tests/gui/attachments/TestAttachmentWidget.cpp:
```diff
@@ -0,0 +1,105 @@
+
+#include <attachments/AttachmentWidget.h>
+
+#include "QTest"
```
`#include <QTest>`

droidmonkey commented on tests/gui/attachments/TestAttachmentWidget.cpp:
```diff
@@ -0,0 +1,105 @@
+#include <qobject.h>
+#include <qt5/QtCore/qglobal.h>
+#include <qtestcase.h>
+#include <qwidget.h>
```
Need to use the non .h includes

droidmonkey commented on tests/gui/attachments/TestAttachmentWidget.cpp:
```diff
@@ -0,0 +1,105 @@
+    QCOMPARE(layout->count(), 0);
+
+    QCOMPARE(layout->alignment(), Qt::AlignCenter);
+}
```
Don't recommend testing the UI itself, tests should be limited to data manipulation and presentation, not ui specifics.

droidmonkey commented on tests/gui/attachments/TestAttachmentWidget.cpp:
```diff
@@ -0,0 +1,105 @@
+    auto label = qobject_cast<QLabel*>(item->widget());
+    QVERIFY(label);
+
+    QCOMPARE(label->text(), tr("Unknown attachment type"));
```
Too specific of a test

w15eacre commented:
@droidmonkey Is MacOs working ok?

droidmonkey commented:
It needs to be fixed, a brew update caused builds to fail

w15eacre committed:
Refactor attachment handling and improve UI text clarity in dialogs and widgets

w15eacre committed:
Refactor attachment handling by simplifying code and improving include statements for clarity

w15eacre committed:
Add null checks for selection model in attachment preview and edit functions

w15eacre commented:
@droidmonkey Does MacOs work?

droidmonkey commented:
Got it working again

droidmonkey commented:
@w15eacre thank you for this PR, great work!
